3ds2-example/backend/src/payments/payments.service.ts

```diff
@@ -0,0 +1,229 @@
+/**
+ * This payments service does NOT use the CheckoutAPI
+ * It calls the API endpoints with axios
+ * (For typing clarity though I am using the models exposed in the node-api-library)
+ */
+
+import { Injectable } from '@nestjs/common';
+import { HttpService } from '@nestjs/axios';
+import { ConfigService } from '@nestjs/config';
+import { catchError, firstValueFrom } from 'rxjs';
+import { AxiosError, AxiosRequestConfig } from 'axios';
+import {
+  PaymentMethodsResponse,
+  PaymentResponse,
+  PaymentRequest,
+  AuthenticationData,
+  ThreeDSRequestData,
+  PaymentDetailsResponse,
+  PaymentCompletionDetails,
+  CreateCheckoutSessionResponse,
+  CreateCheckoutSessionRequest,
+} from '@adyen/api-library/lib/src/typings/checkout/models';
+
+@Injectable()
+export class PaymentsService {
+  API_KEY: string;
+  MERCHANT_ACCOUNT: string;
+  BASE_URL: string;
+
+  constructor(
+    private httpService: HttpService,
+    private configService: ConfigService,
+  ) {
+    this.API_KEY = this.configService.get<string>('API_KEY');
+    this.BASE_URL = this.configService.get<string>('BASE_URL');
+    this.MERCHANT_ACCOUNT = this.configService.get<string>('MERCHANT_ACCOUNT');
+  }
+
+  async postForPaymentMethods(): Promise<PaymentMethodsResponse> {
+    const url = `${this.BASE_URL}/paymentMethods`;
+
+    const headers: AxiosRequestConfig['headers'] = {
+      'x-api-key': this.API_KEY,
+      'Content-Type': 'application/json',
+    };
+    const postData = {
+      merchantAccount: this.MERCHANT_ACCOUNT,
+    };
+
+    const { data } = await firstValueFrom(
+      this.httpService.post<any>(url, postData, { headers }).pipe(
+        catchError((error: AxiosError) => {
+          console.log(error);
+          throw 'An error happened!';
+        }),
+      ),
+    );
+    return data;
+  }
+
+  /**
+   * This method is just a wrapper to route to the correct flow based on the flow parameter which will be NATIVE or REDIRECT
+   * This is an implementation detail of this test app and not related to the 3DS2 flow or integraiton itself.
+   * 
+   * For both native or redirect the more information you pass to the /payments request the higher chance for a frictionless flow (no challenge)
+   * we recommend you at least pass these fields: 
+   *  countryCode: string;
+      shopperName: {
+        firstName: string;
+        lastName: string;
+      },
+      shopperIP: string;
+      shopperEmail: string;
+      Please note: 
+      it is better to omit these fields than pass static values, only populate these with real data from your shopper, otherwise it could result in higher challenge rate
+   */
+  async postForPayments({ data, url, flow }): Promise<PaymentResponse> {
```

Comment:
To show best practices, let's change this so that the frontend is not allowed to pass `flow=='NATIVE'` or `flow=='REDIRECT'` (this should be a backend-only setting).

Can we separate the two functions so it's clear for the user that this is something that backend handles.

_Alternatively: we can add a comment explaining that you're not supposed to do this_ (I prefer the two-functions solution)

Reply:
So I think we need to do this in reverse then, I see your point that currently it's confusing but we can't make this distinction on the backend as the user makes this selection on the client. However, I will change it so that the frontend will explicitly call either native or redirect, so we have two functions on the backend for each flow and encapsulate that implementation detail in the client instead. This will be a lot cleaner.

Reply:
✅